paddle/fluid/framework/ir/ngraph_subgraph_pass.cc

```diff
@@ -0,0 +1,158 @@
+// limitations under the License.
+
+#include <set>
+#include <string>
+#include <unordered_set>
+#include <vector>
+
+#include "paddle/fluid/framework/ir/graph_helper.h"
+#include "paddle/fluid/framework/ir/graph_pattern_detector.h"
+#include "paddle/fluid/framework/ir/ngraph_subgraph_pass.h"
+#include "paddle/fluid/inference/analysis/helper.h"
+#include "paddle/fluid/inference/analysis/ir_passes/subgraph_detector.h"
+#include "paddle/fluid/operators/ngraph/ngraph_bridge.h"
+#include "paddle/fluid/platform/enforce.h"
+#include "paddle/fluid/string/pretty_log.h"
+
+namespace paddle {
+namespace framework {
+namespace ir {
+
+namespace ANAT = paddle::inference::analysis;
+
+std::string GenerateEngineKey(const std::set<std::string> &engine_inputs,
+                              const std::set<std::string> &engine_outputs,
+                              const std::string &size) {
+  std::string engine_hash_key = "";
+  for (auto name : engine_inputs) {
+    engine_hash_key += name;
+  }
+  for (auto name : engine_outputs) {
+    engine_hash_key += name;
+  }
+  engine_hash_key += size;
+  auto engine_key = std::to_string(std::hash<std::string>()(engine_hash_key));
+  return engine_key;
+}
+
+void NgraphSubgraphPass::ApplyImpl(ir::Graph *graph) const {
+  PADDLE_ENFORCE(graph);
+  FusePassBase::Init("ngraph_subgraph_pass", graph);
+
+  std::unordered_set<Node *> nodes2delete;
+
+  auto teller = [](const Node *node) {
+    if (!node->IsOp() || !node->Op()) return false;
+    auto op_type = node->Op()->Type();
+    return !paddle::operators::NgraphBridge::isRegister(op_type);
+  };
+
+  ANAT::SubGraphFuser fuser(graph, teller, 0, "ngraph_engine");
+  fuser();
+
+  for (auto *node : graph->Nodes()) {
+    if (node->IsOp() && !ANAT::Agent(node).subgraph()->empty()) {
+      OpDesc *op_desc = node->Op();
+      op_desc->SetType("ngraph_engine");
+      for (auto it = ANAT::Agent(node).subgraph()->begin();
+           it != ANAT::Agent(node).subgraph()->end(); ++it) {
+      }
+
+      CreateNgraphEngineOp(node, graph);
+
+      std::unordered_set<const Node *> nodes2remove(
+          ANAT::Agent(node).subgraph()->begin(),
+          ANAT::Agent(node).subgraph()->end());
+      GraphSafeRemoveNodes(graph, nodes2remove);
+    }
+  }
+
+  std::unordered_set<const Node *> nodes2remove;
+  for (auto *node : graph->Nodes()) {
+    if (node->IsOp() && ANAT::Agent(node).deleted()) {
+      nodes2remove.insert(node);
+    }
+  }
+  framework::ir::GraphSafeRemoveNodes(graph, nodes2remove);
+  std::vector<ir::Node *> nodes = ir::TopologySortOperations(*graph);
+}
+
+void NgraphSubgraphPass::CreateNgraphEngineOp(framework::ir::Node *node,
+                                              Graph *graph) const {
+  auto *op_desc = node->Op();
+  auto &subgraph = *ANAT::Agent(node).subgraph();
+  PADDLE_ENFORCE(!subgraph.empty());
+
+  framework::ProgramDesc *program_desc =
+      Get<framework::ProgramDesc *>("program");
```

Comment:
Is that fine to us directly modify the program desc @Superjomn @luotao1

Reply:
I was following tensorrt subgraph approach. https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/inference/analysis/ir_passes/tensorrt_subgraph_pass.cc#L97